src/pmdas/farm/pmda.c

```diff
@@ -0,0 +1,1907 @@
+		return sts;
+
+	return pmdaFetch(numpmid, pmidlist, resp, pmda);
+}
+
+static int
+farm_fetchCallBack(pmdaMetric *mdesc, unsigned int inst, pmAtomValue *atom)
+{
+	unsigned int 		item = pmID_item(mdesc->m_desc.pmid);
+	unsigned int		cluster = pmID_cluster(mdesc->m_desc.pmid);
+	struct seagate_disk 	*dev;
+	int 			sts;
+
+	switch (cluster) {
+		case CLUSTER_ATA_LOG_HEADER:
+		case CLUSTER_ATA_DRIVE_INFORMATION:
+		case CLUSTER_ATA_WORKLOAD_STATISTICS:
+		case CLUSTER_ATA_ERROR_STATISTICS:
+		case CLUSTER_ATA_ENVIRONMENTAL_STATISTICS:
+		case CLUSTER_ATA_RELIABILITY_STATISTICS:
+			sts = pmdaCacheLookup(INDOM(DISK_INDOM), inst, NULL, (void **)&dev);
+			if (sts < 0)
+				return sts;
+			return farm_ata_data_fetch(item, cluster, &dev->farm_ata_log_stats, atom);
+
+		default:
+			return PM_ERR_PMID;
+	}
+
+	return PMDA_FETCH_STATIC;
+}
+
+void
+farm_instance_setup(void)
+{
+	static char lsblk_command[] = "lsblk -d -n -e 1,2,7,11,252 -o name";
+	static char smart_command[] = "LC_ALL=C smartctl";
+	char *env_command;
+
+	/* allow override at startup for QA testing */
+	if ((env_command = getenv("FARM_SETUP_LSBLK")) != NULL)
+		farm_setup_lsblk = env_command;
+	else
+		farm_setup_lsblk = lsblk_command;
+		
+	/* allow override at startup for QA testing */
+	if ((env_command = getenv("FARM_SETUP_SMARTCTL")) != NULL)
+		farm_setup_smartctl = env_command;
+	else
+		farm_setup_smartctl = smart_command;
+
+}
+
+static int
+farm_text(int ident, int type, char **buf, pmdaExt *pmda)
+{
+	if ((type & PM_TEXT_PMID) == PM_TEXT_PMID) {
+		int sts = pmdaDynamicLookupText(ident, type, buf, pmda);
+		if (sts != -ENOENT)
+			return sts;
+	}
+	return pmdaText(ident, type, buf, pmda);
+}
+
+static int
+farm_pmid(const char *name, pmID *pmid, pmdaExt *pmda)
+{
+	pmdaNameSpace *tree = pmdaDynamicLookupName(pmda, name);
+	return pmdaTreePMID(tree, name, pmid);
+}
+
+static int
+farm_name(pmID pmid, char ***nameset, pmdaExt *pmda)
+{
+	pmdaNameSpace *tree = pmdaDynamicLookupPMID(pmda, pmid);
+	return pmdaTreeName(tree, pmid, nameset);
+}
+
+static int
+farm_children(const char *name, int flag, char ***kids, int **sts, pmdaExt *pmda)
+{
+	pmdaNameSpace *tree = pmdaDynamicLookupName(pmda, name);
+	return pmdaTreeChildren(tree, name, flag, kids, sts);
+}
+
+void
+__PMDA_INIT_CALL
+farm_init(pmdaInterface *dp)
+{
+	int nindoms = sizeof(indomtable)/sizeof(indomtable[0]);
+	int nmetrics = sizeof(metrictable)/sizeof(metrictable[0]);
+
+	if (_isDSO) {
+		char helppath[MAXPATHLEN];
+		int sep = pmPathSeparator();
+		pmsprintf(helppath, sizeof(helppath), "%s%c" "farm" "%c" "help",
+			pmGetConfig("PCP_PMDAS_DIR"), sep, sep);
+		pmdaDSO(dp, PMDA_INTERFACE_4, "FARM DSO", helppath);
+	}
+
+	if (dp->status != 0)
+		return;
+
+	/* Check for environment variables allowing test injection */
+	farm_instance_setup();
+	farm_stats_setup();
+
+	dp->version.four.instance = farm_instance;
+	dp->version.four.fetch = farm_fetch;
+	dp->version.four.text = farm_text;
```

Comment:
We don't use dynamic metrics in this PMDA, so I don't think we need farm_text, farm_pmid, farm_name or farm_children callbacks, the libpcp_pmda defaults should do the trick.  If this was based on the Linux agent (?) this is a historical quirk - should be removed there too I think (I'll look into that) - we used to have dynamic interrupts metrics there, but no longer do.

We'll need new label callbacks instead - 
```
    dp->version.seven.label = farm_label;
    pmdaSetLabelCallBack(dp, farm_labelCallBack);
```

See the Linux agent for a reference and pmLookupLabels(3).